File: extra/packFonts.m
function packFonts()
% run extra/genfont.sh first

fonts = dir(fullfile('extra','fonts')) ;
fonts = {fonts([fonts.isdir]).name} ;
fonts(ismember(fonts,{'.','..'})) = [] ;
chars = 'a':'z' ;

im = cell(numel(fonts), numel(chars)) ;
labels = cell(numel(fonts), numel(chars)) ;

for i = 1:numel(fonts)
  for j = 1:numel(chars)
    [p,cmap] = imread(fullfile('extra', 'fonts', fonts{i}, [chars(j) '.png'])) ;
    im{i,j} = im2single(ind2gray(p,cmap)) ;
    labels{i,j} = j ;
  end
end

imdb.meta.classes = chars ;
imdb.meta.sets = {'train', 'val'} ;
imdb.meta.fonts = fonts ;
imdb.images.id = 1:numel(im) ;
imdb.images.data = cat(3, im{:}) ;
imdb.images.label = cat(2, labels{:}) ;

% Create training and validation sets
sets = ones(numel(fonts), numel(chars)) ;
sets(end-349:end,:) = 2 ;
imdb.images.set = sets(:)' ;

save('data/charsdb.mat', '-struct', 'imdb') ;

File: extra/pack_fonts.py
# run extra/genfont.sh first
import os
import torch
import lab

font_dir = os.path.join('extra','fonts')
fonts = [name for name in os.listdir(font_dir) if
    os.path.isdir(os.path.join(font_dir, name))]
fonts = sorted(fonts)

chars = [chr(c) for c in range(ord('a'),ord('z')+1)]
images = []
labels = []

for f, font in enumerate(fonts):
    for c, char in enumerate(chars):
        im_file = os.path.join(font_dir, font, char + '.png')
        im = lab.imread(im_file)
        images.append(im)
        labels.append(c)
        print(f"Added {im_file}")

images = torch.cat(images,0)
labels = torch.tensor(labels)
sets = torch.zeros(len(fonts), len(chars), dtype=torch.int32)
sets[-349:,:] = 1
sets = sets.reshape(-1)

imdb = {
    'meta' : {
        'classes' : chars,
        'sets' : ['train', 'val'],
        'fonts' : fonts,
    },
    'images' : images,
    'labels' : labels,
    'sets' : sets,
}
torch.save(imdb, os.path.join('data', 'charsdb.pth'))

